sycl/plugins/cuda/pi_cuda.cpp

```diff
@@ -67,19 +67,40 @@ inline void assign_result(pi_result *ptr, pi_result value) noexcept {
 // Invokes the callback for each event in the wait list. The callback must take
 // a single pi_event argument and return a pi_result.
 template <typename Func>
-pi_result forEachEvent(const pi_event *event_wait_list,
-                       std::size_t num_events_in_wait_list, Func &&f) {
+pi_result forLatestEvents(const pi_event *event_wait_list,
+                          std::size_t num_events_in_wait_list, Func &&f) {
 
   if (event_wait_list == nullptr || num_events_in_wait_list == 0) {
     return PI_INVALID_EVENT_WAIT_LIST;
   }
 
-  for (size_t i = 0; i < num_events_in_wait_list; i++) {
-    auto event = event_wait_list[i];
-    if (event == nullptr) {
-      return PI_INVALID_EVENT_WAIT_LIST;
+  // Fast path if we only have a single event
+  if (num_events_in_wait_list == 1) {
+    return f(event_wait_list[0]);
+  }
+
+  std::vector<pi_event> events{event_wait_list,
+                               event_wait_list + num_events_in_wait_list};
+  std::sort(events.begin(), events.end(), [](pi_event e0, pi_event e1) {
```

Comment:
wouldn't it be more effective to keep track of the last event in the stream as they are created?

Reply:
The latest event in a relevant stream is not necessarily in the event list, so we may lose performance from waiting for it.

Consider for example that we have 3 events in a queue and want to synchronize host with the first two events, but not the last. Since events are in order we could wait for the second event only, but if we waited for the last event we may be wasting the time between the second and the last event unnecessarily.

Reply:
This makes sense, thanks.